eacl/validator.go

```diff
@@ -0,0 +1,169 @@
+package eacl
+
+import (
+	"bytes"
+	"errors"
+
+	"github.com/nspcc-dev/neofs-api-go/pkg/acl/eacl"
+	"go.uber.org/zap"
+)
+
+// Validator is a tool that calculates
+// the action on a request according
+// to the extended ACL rule table.
+type Validator struct {
+	*cfg
+}
+
+// NewValidator creates and initializes a new Validator using options.
+func NewValidator(opts ...Option) *Validator {
+	cfg := defaultCfg()
+
+	for i := range opts {
+		opts[i](cfg)
+	}
+
+	return &Validator{
+		cfg: cfg,
+	}
+}
+
+// CalculateAction calculates action on the request according
+// to its information represented in ValidationUnit.
+//
+// The action is calculated according to the application of
+// eACL table of rules to the request.
+//
+// If the eACL table is not available at the time of the call,
+// eacl.ActionUnknown is returned.
+//
+// If no matching table entry is found, ActionAllow is returned.
+func (v *Validator) CalculateAction(unit *ValidationUnit) eacl.Action {
+	var (
+		err   error
+		table *eacl.Table
+	)
+
+	if unit.bearer != nil {
+		table = eacl.NewTableFromV2(unit.bearer.GetBody().GetEACL())
+	} else {
+		// get eACL table by container ID
+		table, err = v.storage.GetEACL(unit.cid)
+		if err != nil {
+			if errors.Is(err, ErrEACLNotFound) {
```

Comment:
Local defined error shouldn't be expected from external component, I suppose. In fact this condition is never met in neofs-node right now, because container contract wrapper returns `container.ErrEACLNotFound`. Then validator returns `eacl.ActionUnknown` and request fails.

I am not sure what the best solution for this:
- return (action, error) and process error type outside of the validator
- do not expect error from `v.storage` if eACL is not set,
- ...

Reply:
https://github.com/nspcc-dev/neofs-node/pull/201
Here container wrapper started to return errors in case of empty eACL.

Reply:
That's really a part of the interface definition, so it makes sense to define the error right here.
>return (action,error)
I like the idea. I am now thinking about accepting table in `CalculateUnit` and fetching it from store on the caller side. Coupling made sense when everything was in a single repo, SDK should be more granular. This will allow to remove both an error and `Source` intreface.

Having both table and err `== nil` is a bit risky. We may extend this to `table, ok, err :=`, though.

Reply:
Fixed.